services/gundeck/src/Gundeck/Push/Websocket.hs

```diff
@@ -1,37 +1,280 @@
+    groupByURI = groupAssoc . fmap (\(notif, (uri, prc)) -> (uri, (notif, prc)))
+
+    pullUri :: (notif, [Presence]) -> [(notif, (URI, Presence))]
+    pullUri (notif, prcs) = (notif,) . (bulkresource &&& id) <$> prcs
+
+bulkSend :: URI -> BulkPushRequest -> Gundeck (URI, Either SomeException BulkPushResponse)
+bulkSend uri req = (uri,) <$> ((Right <$> bulkSend' uri req) `catch` (pure . Left))
+
+bulkSend' :: URI -> BulkPushRequest -> Gundeck BulkPushResponse
+bulkSend' uri (encode -> jsbody) = do
+    req <- ( check
+           . method POST
+           . contentJson
+           . lbytes jsbody
+           . timeout 3000 -- ms
+           ) <$> Http.setUri empty (fromURI uri)
+    try (submit req) >>= \case
+        Left  e -> throwM (e :: SomeException)
+        Right r -> decodeBulkResp $ responseBody r
+  where
+    submit req = recovering (limitRetries 1) rpcHandlers $ const (rpc' "cannon" req id)
+
+    check req = req { Http.checkResponse = \rq rs ->
+        when (responseStatus rs /= status200) $
+            let ex = StatusCodeException (rs { responseBody = () }) mempty
+            in throwM $ HttpExceptionRequest rq ex
+    }
+
+    decodeBulkResp :: MonadThrow m => Maybe L.ByteString -> m BulkPushResponse
+    decodeBulkResp Nothing    = throwM $ ErrorCall "missing response body from cannon"
+    decodeBulkResp (Just lbs) = either err pure $ eitherDecode lbs
+      where err = throwM . ErrorCall . ("bad response body from cannon: " <>)
+
+-- | NOTE: 'PushTarget's may occur several times both in the "lost" and in the "delivered" list.
+-- This happens iff there are several 'Notifcation's for the same 'PushTarget', and some of them are
+-- delivered while others aren't.
+data FlowBack = FlowBack
+    { flowBackBadCannons :: [(URI, SomeException)]  -- ^ list of cannons that failed to respond with status 200
+    , flowBackLostPrcs   :: [PushTarget]            -- ^ 401 inside the body (for one presence)
+    , flowBackDelivered  :: [(NotificationId, PushTarget)]
+    }
+
+flowBack :: [(URI, Either SomeException BulkPushResponse)] -> FlowBack
+flowBack rawresps = FlowBack broken gone delivered
+  where
+    broken :: [(URI, SomeException)]
+    broken
+        = catLefts rawresps
+
+    gone :: [PushTarget]
+    gone
+        = map (snd . snd)
+        . filter (\(st, _) -> case st of
+                     PushStatusOk   -> False
+                     PushStatusGone -> True)
+        $ responsive
+
+    delivered :: [(NotificationId, PushTarget)]
+    delivered
+        = map snd
+        . filter (\(st, _) -> case st of
+                     PushStatusOk   -> True
+                     PushStatusGone -> False)
+        $ responsive
+
+    responsive :: [(PushStatus, (NotificationId, PushTarget))]
+    responsive = map (\(n, t, s) -> (s, (n, t)))
+               . mconcat . fmap fromBulkPushResponse . catRights $ snd <$> rawresps
+
+    catRights :: [Either a b] -> [b]
```

Comment:
Any reason not to use `rights` from `Data.Either`?

Reply:
yes: didn't think of it.  thanks!  (-: